graphql_client_codegen/src/lib.rs

```diff
@@ -44,67 +43,92 @@ impl std::error::Error for GeneralError {}
 
 type BoxError = Box<dyn std::error::Error + Send + Sync + 'static>;
 type CacheMap<T> = std::sync::Mutex<BTreeMap<std::path::PathBuf, T>>;
+type QueryDocument = graphql_parser::query::Document<'static, String>;
 
 lazy_static! {
-    static ref SCHEMA_CACHE: CacheMap<schema::Schema> = CacheMap::default();
-    static ref QUERY_CACHE: CacheMap<(String, graphql_parser::query::Document<'static, String>)> =
-        CacheMap::default();
+    static ref SCHEMA_CACHE: CacheMap<Schema> = CacheMap::default();
+    static ref QUERY_CACHE: CacheMap<(String, QueryDocument)> = CacheMap::default();
 }
 
-/// Generates Rust code given a query document, a schema and options.
+fn get_set_cached<T: Clone>(
+    cache: &CacheMap<T>,
+    key: &std::path::Path,
+    value_func: impl FnOnce() -> T,
+) -> T {
+    let mut lock = cache.lock().expect("cache is poisoned");
+    lock.entry(key.into()).or_insert_with(value_func).clone()
+}
+
+fn query_document(query_string: &str) -> Result<QueryDocument, BoxError> {
+    let document = graphql_parser::parse_query(query_string)
+        .map_err(|err| GeneralError(format!("Query parser error: {}", err)))?
+        .into_static();
+    Ok(document)
+}
+
+fn get_set_query_from_file(query_path: &std::path::Path) -> (String, QueryDocument) {
+    get_set_cached(&QUERY_CACHE, query_path, || {
+        let query_string = read_file(query_path).unwrap();
+        let query_document = query_document(&query_string).unwrap();
+        (query_string, query_document)
+    })
+}
+
+fn get_set_schema_from_file(schema_path: &std::path::Path) -> Schema {
+    get_set_cached(&SCHEMA_CACHE, schema_path, || {
+        let schema_extension = schema_path
+            .extension()
+            .map(|ext| ext.to_str().expect("Path must be valid UTF-8"))
+            .unwrap_or("<no extension>");
+        let schema_string = read_file(schema_path).unwrap();
+        match schema_extension {
+            "graphql" | "gql" => {
```

Comment:
Hello, I think we should also support `graphqls` as it is also a valid schema file extensions